airflowctl: initialize extra for create assets event

sidshas03 opened the pull request:
Align airflowctl with the FastAPI contract for **POST `/assets/events`**.

**What**
- Use absolute path `"/assets/events"` in `AssetsOperations.create_event` (was `"assets/events"`).
- Add a small client test asserting route + payload:
  - `airflow-ctl/tests/test_create_event_client.py`
- Request/response match the datamodel (`CreateAssetEventsBody` → `AssetEventResponse`).

**Why**
- Avoid routing issues from relative paths.
- Keep `airflowctl` in sync with API + models (pairs with the POST `/assets` work).

Related: #55697
Companion: #55702

**Verify**
```bash
python -m pytest -q airflow-ctl/tests -k create_event

bugraoz93 commented on airflow-ctl/tests/test_create_event_client.py:
```diff
@@ -0,0 +1,36 @@
+import json
```
I don't think we need this test, we already have unit tests here
https://github.com/apache/airflow/blob/1576db7baaee2890727ae3e0fda33eafce4e5e31/airflow-ctl/tests/airflow_ctl/api/test_operations.py#L334

I think this hasn't been caught because of it needs integration tests

sidshas03 replied:
Ack, removed. We already cover create_event in airflow_ctl/api/test_operations.py. I’ll add integration coverage in the follow up PR.

bugraoz93 commented on airflow-ctl/src/airflowctl/api/operations.py:
```diff
@@ -238,13 +239,23 @@ def list_by_alias(self) -> AssetAliasCollectionResponse | ServerResponseError:
         try:
             self.response = self.client.post(
-                "assets/events", json=_date_safe_dict_from_pydantic(asset_event_body)
+                "/assets/events", json=_date_safe_dict_from_pydantic(asset_event_body)
```
Maybe we only left this part in this PR and keep the create part here https://github.com/apache/airflow/blob/1576db7baaee2890727ae3e0fda33eafce4e5e31/airflow-ctl/tests/airflow_ctl/api/test_operations.py#L334

We can merge them separately and make it easier to review.  We only need this line for the scope of the ticket so that this can quickly merged as fix

sidshas03 replied:
Done! left only the "/assets/events" change here; everything else will go into the follow-up PR.

bugraoz93 reviewed (commented):
Thanks a lot for the fix! I think we can separate and merge the fix quicker :)

bugraoz93 reviewed (approved):
Thanks for the update! That should be enough. No need to include tests, we already have tests here https://github.com/apache/airflow/blob/1576db7baaee2890727ae3e0fda33eafce4e5e31/airflow-ctl/tests/airflow_ctl/api/test_operations.py#L334. What is happening for the bug couldn't be caught with the unit tests. We will implement integration tests in another issue soon. That's why for the test part, we don't need to follow up on unit test side

bugraoz93 reviewed (changes requested):
```
(airflow) bugra@dev:~/PycharmProjects/airflow$ airflowctl assets create-event --asset-id=1
2025-09-17 22:52:40 [warning  ] Server error                   [airflowctl.api.client] extra={'detail': [{'type': 'dict_type', 'loc': ['body', 'extra'], 'msg': 'Input should be a valid dictionary', 'input': None}]}
Server response error: Client error message: {'detail': [{'type': 'dict_type', 'loc': ['body', 'extra'], 'msg': 'Input should be a valid dictionary', 'input': None}]}
Client error,  Please check the command and its parameters. If you need help, run the command with --help.
```
I don't think this resolves the bug. I have tested locally, but I still get the above error. I think that part can stay without `/`. If you check the entire file, config is the only one that has the `/` at the beginning. I think that doesn't follow the convention, even though it is working (I believe it is handled by the client). So I think it really doesn't change anything to have the `/` there 
You can test your changes by running Breeze and the API server. You can also use airflowctl with `uv sync` and test it locally.

bugraoz93 commented on airflow-ctl/src/airflowctl/api/operations.py:
```diff
@@ -244,7 +244,7 @@ def create_event(
         """Create an asset event."""
         try:
             self.response = self.client.post(
```
```suggestion
            if asset_event_body.extra is None:
                asset_event_body.extra = {}
            self.response = self.client.post(
```
I think this will fix the bug :) This is some kind of integration problem we are having with dictionary fields in airflowctl (There are a couple of cases we have dict extra fields, such as in `connections`.  In some cases, in the database, they cannot be `null`, but it can be `null` on the datamodel, which ends up in error if it is passed as `None`
Please see the example below; it is the same problem. Let's patch it like this for now and we can come up with more generic solution to this

https://github.com/apache/airflow/blob/8d891bb98f318a15b063d9e9faa4a4fd6b49ef3d/airflow-ctl/src/airflowctl/api/operations.py#L573-L585

sidshas03 replied:
Applied the suggested fix in c3555b0 extra now defaults to {} when None; kept the absolute path. Tested locally and it works. Re-requesting review.

sidshas03 committed:
airflowctl: ensure event.extra defaults to {} when None

- Add guard in create_event method to ensure extra is never null
- Change path from 'assets/events' to '/assets/events' for consistency
- Prevents DB/UI issues when extra field is None

bugraoz93 commented on airflow-ctl/src/airflowctl/api/operations.py:
```diff
@@ -243,8 +243,11 @@ def create_event(
         """Create an asset event."""
         try:
+            # Ensure extra is an object; DB/UI choke on null
+            if getattr(asset_event_body, "extra", None) is None:
```
```suggestion
            if asset_event_body.extra is None:
```

Even though I like `getattr` like safe calls, this actually brings danger to not fail the unit test. If `extra` is removed or changed name is changed in the datamodel (Pydantic models), it creates a bug. Because this statement will return None in that case and we always add extra even though the API already changed it. This is why we use Pydantic which is kind of a data contract between the API and the client 

You can try the behavior even with a dict 
```python
a={}
if getattr(a, "extra", None) is None:
    a["extra"] = {}
    print(a)
```

sidshas03 replied:
Updated as suggested in **aeef5f7** — switched to a direct `asset_event_body.extra is None` check so we don’t mask schema drift; kept the absolute `"/assets/events"` path. Verified locally. Ready for re-review @bugraoz93  thanks!

sidshas03 committed:
airflowctl: use direct attribute check for event.extra (avoid getattr masking schema changes)

bugraoz93 commented on airflow-ctl/src/airflowctl/api/operations.py:
```diff
@@ -243,8 +243,11 @@ def create_event(
+                asset_event_body.extra = {}
             self.response = self.client.post(
-                "assets/events", json=_date_safe_dict_from_pydantic(asset_event_body)
+                "/assets/events", json=_date_safe_dict_from_pydantic(asset_event_body)
```
I mentioned a couple of times that this should not actually align with multiple of them, and we should use the other way around. That part is working. We don't need to update it
```suggestion
                "assets/events", json=_date_safe_dict_from_pydantic(asset_event_body)
```

bugraoz93 commented on airflow-ctl/src/airflowctl/api/operations.py:
```diff
@@ -243,8 +243,11 @@ def create_event(
     ) -> AssetEventResponse | ServerResponseError:
         """Create an asset event."""
         try:
+            # Ensure extra is an object; DB/UI choke on null
```
There is no `choke` on

```suggestion
            # Ensure extra is initialised before sent to API
```

bugraoz93 committed:
Update airflow-ctl/src/airflowctl/api/operations.py

bugraoz93 committed:
Update airflow-ctl/src/airflowctl/api/operations.py
